extensions/avro/src/main/java/com/hazelcast/jet/avro/AvroSerializerHooks.java

```diff
@@ -64,4 +179,216 @@ public boolean isOverwritable() {
             return false;
         }
     }
+
+    private abstract static class LazyImmutableContainer<T extends GenericContainer> implements GenericContainer {
+        private final byte[] serialized;
+        private final Schema schema;
+        private T deserialized;
+
+        protected LazyImmutableContainer(byte[] serialized, Schema schema) {
+            this.serialized = serialized;
+            this.schema = schema;
+        }
+
+        protected T deserialized() {
+            if (deserialized == null) {
+                deserialized = deserialize(new GenericDatumReader<>(schema), serialized);
```

Comment:
if no thread safety/single deserialization is needed, add a comment

Reply:
12105f11. `deserialized()` is idempotent and, I think, concurrent initializations can be ignored.